### PR TITLE
Add eleveldb:close 

### PR DESCRIPTION
**Pull request by jonmeredith**
Added to fix the missing MANIFEST race around handoff / destroy / open.

This adds a mutex to the db_handle object to make sure it had not been closed from underneath any operations.  All iterators created from the DB Ref are tracked and cleaned up when the database is closed.  Any folds will throw exceptions if in progress and any other operation should return {error, einval} (modelled after the socket library).

As well as adding close, this also fixes a use-after-free bug to do with garbage collection.  Although the NIF resources are released in the previous code, the order that the cleanup run is determined by the position in the heap by the garbage collector.  The two examples show different ordering of cleanup.

```

11> spawn(fun() -> {ok, Db} = eleveldb:open("/tmp/t.db",[{create_if_missing, true}]), eleveldb:put(Db, <<"k">>,<<"v">>,[]), eleveldb:fold(Db, fun(_,_) -> lists:seq(1,10000) end, undefined, []) end).
**** db 0x7f8b220fda28 cleanup
**** iterator 0x1e261ae8 cleanup for db 0x7f8b220fda28
<0.54.0>
12> spawn(fun() -> {ok, Db} = eleveldb:open("/tmp/t.db",[{create_if_missing, true}]), eleveldb:put(Db, <<"k">>,<<"v">>,[]), eleveldb:fold(Db, fun(_,_) -> lists:seq(1,10000) end, undefined, []), garbage_collect() end).
**** iterator 0x7f8b220fca58 cleanup for db 0x7f8b220fc9c0
**** db 0x7f8b220fc9c0 cleanup
```

The code needs to be benchmarked before merging which we'll work on, but can be reviewed before then.

### CODE REVIEW

File: c_src/eleveldb.cc [hunk 1]
```diff
@@ -251,6 +260,61 @@ ERL_NIF_TERM write_batch_item(ErlNifEnv* env, ERL_NIF_TERM item, leveldb::WriteB
...
+             ++iters_it)
+        {
+            eleveldb_itr_handle* itr_handle = *iters_it;
+            enif_mutex_lock(itr_handle->itr_lock);
+            free_itr(*iters_it);
+            enif_mutex_unlock(itr_handle->itr_lock);
+        }
+
+        // close the db 
+        if (db_handle->db)
```
Comment (jonmeredith) [inline c_src/eleveldb.cc]: No need to recheck - already know db_handle->db != NULL

File: c_src/eleveldb.cc [hunk 2]
```diff
@@ -284,7 +348,9 @@ ERL_NIF_TERM eleveldb_open(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
                                                        sizeof(eleveldb_db_handle));
         memset(handle, '\0', sizeof(eleveldb_db_handle));
         handle->db = db;
+        handle->db_lock = enif_mutex_create((char*)"eleveldb_db_lock");
         handle->options = opts;
+        handle->iters = std::set<struct eleveldb_itr_handle*>();
```
Comment (jonmeredith) [inline c_src/eleveldb.cc]: Check with C++ expert this is a good way to initialize a class in allocated memory

File: c_src/eleveldb.cc [hunk 3]
```diff
@@ -496,17 +617,17 @@ ERL_NIF_TERM eleveldb_iterator_close(ErlNifEnv* env, int argc, const ERL_NIF_TER
     eleveldb_itr_handle* itr_handle;
     if (enif_get_resource(env, argv[0], eleveldb_itr_RESOURCE, (void**)&itr_handle))
     {
+        enif_mutex_lock(itr_handle->db_handle->db_lock);
```
Comment (jonmeredith) [inline c_src/eleveldb.cc:623]: Add comment to explain acquiring locks in the same order.

File: c_src/eleveldb.cc [hunk 4]
```diff
@@ -645,7 +774,14 @@ static void eleveldb_itr_resource_cleanup(ErlNifEnv* env, void* arg)
     {
         delete itr_handle->itr;
```
Comment (jonmeredith) [inline c_src/eleveldb.cc]: call free_itr instead of cleanup code

File: c_src/eleveldb.cc (hunk 1 above)
Comment (jonmeredith) [inline c_src/eleveldb.cc]: Fixed.

File: c_src/eleveldb.cc (hunk 3 above)
Comment (jonmeredith) [inline c_src/eleveldb.cc:623]: Fixed.

**Comment by slfritchie** [comment]
+1 to merge.  After chatting with Jon, my eyes think that this looks sane enough.

**Comment by freerobby** [comment]
Thanks for working on this. We got stung by this race condition this morning and it took down two nodes in our cluster. I'll likely upgrade to a 1.2 release candidate as soon as this is merged in.

**Comment by jonmeredith** [comment]
@freerobby what platform/version were you running and how exactly did it fail?  This definitely resolves a missing MANIFEST around handoff issue.  We've had some coredumps related to memory management, but haven't been able to definitely pin it on this issue.

**Comment by freerobby** [comment]
@jonmeredith We're running riak 1.1.2 on Ubuntu 12.04.

This was the top of our stack trace:

2012-07-18 17:28:47.287 [error] <0.560.0>@riak_kv_vnode:init:236 Failed to start riak_kv_eleveldb_backend Reason: {db_open,"IO error: /mnt/riak/leveldb/182687704666362864775460604089535377456991567872/MANIFEST-000002: No such file or directory"}

I checked by hand and verified the mentioned file was not on our file system. I was able to fix the node by deleting the /mnt/riak/leveldb/182687704666362864775460604089535377456991567872 directory. I suspect there is a better quick solution but I wanted to get the node back online quickly. All of our data has n=3 and we had a backup that was only one hour old, so I wasn't too concerned about data loss.

**Comment by jonmeredith** [comment]
ok, looks very familiar.  there should be a handoff entry for 182687704666362864775460604089535377456991567872 somewhat recently before the crash.

**Comment by freerobby** [comment]
Yep, indeed. Check the timestamps and partition IDs:

From console.log:

2012-07-18 08:13:26.298 [info] <0.29244.159>@riak_core_handoff_sender:start_fold:144 Handoff of partition riak_kv_vnode 182687704666362864775460604089535377456991567872 from 'riak@10.123.46.124' to 'riak@10.86.199.20' completed: sent 271350 objects in 269.95 seconds

From error.log:

2012-07-18 08:13:26.715 [error] <0.31447.159>@riak_kv_vnode:init:236 Failed to start riak_kv_eleveldb_backend Reason: {db_open,"IO error: /mnt/riak/leveldb/182687704666362864775460604089535377456991567872/MANIFEST-000002: No such file or directory"}

**Comment by jaredmorrow** [comment]
@freerobby this change will make it into Riak 1.2.0 RC2 that will be out Friday 07/20

**Pull request merged**